README.md
# (MODX)EvolutionCMS.plugins.ManagerManager.mm_minimizableSections

A widget for ManagerManager plugin that allows sections to be minimizable on the document edit page.


## Requires

* PHP >= 5.4
* [(MODX)EvolutionCMS.plugins.ManagerManager](https://code.divandesign.ru/modx/managermanager) >= 0.7


## Installation

To install you must unzip the archive to `/assets/plungins/managermanager/widgets/`.


You may also read this documentation:
* [(MODX)EvolutionCMS.plugins.ManagerManager](https://code.divandesign.ru/modx/managermanager).
* [(MODX)EvolutionCMS.modules.ddMMEditor](https://code.divandesign.ru/modx/ddmmeditor).


## Parameters description

* `$params`
	* Desctription: The object of parameters.
	* Valid values:
		* `stdClass`
		* `arrayAssociative`
	* Default value: —
	
* `$params->sections`
	* Desctription: The ID(s) of the sections that the widget is applied to.
	* Valid values:
		* `''` — all sections
		* `stringCommaSeparated`
		* `array`
	* Default value: `''`
	
* `$params->minimizedByDefault`
	* Desctription: The ID(s) of the sections that should be minimized by default.
	* Valid values:
		* `stringCommaSeparated`
		* `array`
	* Default value: —
	
* `$params->roles`
	* Desctription: The CMS user roles that the widget is applied to.
	* Valid values:
		* `array`
		* `stringCommaSeparated`
		* `''` — when this parameter is empty then widget is applied to the all roles
	* Default value: `''`
	
* `$params->roles[$i]`
	* Desctription: CMS user role ID.
	* Valid values: `integer`
	* **Required**
	
* `$params->templates`
	* Desctription: Document templates IDs for which the widget is applied to.
	* Valid values:
		* `array`
		* `stringCommaSeparated`
		* `''` — empty value means the widget is applying to all templates
	* Default value: `''`
	
* `$params->templates[$i]`
	* Desctription: Template ID.
	* Valid values: `integer`
	* **Required**


## CMS events

* `OnDocFormPrerender`
* `OnDocFormRender`


## Examples


### Apply to all sections for all roles and templates

```php
mm_minimizableSections();
```


### Apply to the content and Template Variables sections for users with role `1` editing the documents with the template ID equals `3`

```php
mm_minimizableSections([
	'sections' => 'content,tvs',
	'roles' => '1',
	'templates' => '3',
]);
```

### Apply to all sections and set minimized Template variables and photos sections by default

```php
mm_minimizableSections([
	'minimizedByDefault' => 'tvs,photos',
]);
```


## Links

* [Home page](https://code.divandesign.ru/modx/mm_minimizablesections)
* [Telegram chat](https://t.me/dd_code)
* [Packagist](https://packagist.org/packages/dd/evolutioncms-plugins-managermanager-mm_minimizablesections)
* [GitHub](https://github.com/DivanDesign/EvolutionCMS.plugins.ManagerManager.mm_minimizableSections)


<link rel="stylesheet" type="text/css" href="https://raw.githack.com/DivanDesign/CSS.ddMarkdown/master/style.min.css" />

CHANGELOG.md
# (MODX)EvolutionCMS.plugins.ManagerManager.mm_minimizableSections changelog


## Version 0.3 (2016-11-11)
* \* Parameters: The widget now uses “named” parameters, see README → Examples (with backward compatibility).
* \* Parameters → `$params->sections`:
	* \* Valid values → `''`: Empty value means “for all sections” (with backward compatibility).
	* \* Default value: The parameter is not required anymore and equals to `''` by default.
* \* Attention! PHP >= 5.4 is required.
* \* Attention! (MODX)EvolutionCMS.plugins.ManagerManager >= 0.7 is required.


## Version 0.1 (2015-05-30)
* \+ The first release.


<link rel="stylesheet" type="text/css" href="https://raw.githack.com/DivanDesign/CSS.ddMarkdown/master/style.min.css" />
<style>ul{list-style:none;}</style>

jQuery.ddMM.mm_minimizableSections.js
/**
 * jQuery.ddMM.mm_minimizableSections
 * @version 1.1.1 (2023-06-06)
 * 
 * @copyright 2016–2023
 */

(function($){
$.ddMM.mm_minimizableSections = {
	defaults: {
		classNames: {
			header: 'mm_minimizableSections_header',
			minimized: 'mm_minimizableSections_minimized'
		}
	},
	
	inited: false,
	
	/**
	 * @method init
	 * @version 1.0.1 (2023-06-05)
	 * 
	 * @desc Initialization.
	 * 
	 * @param params {object_plain} — Parameters passed as plain object.
	 * @param params.minimizedByDefault {string} — Minimized by default sections.
	 * 
	 * @returns {void}
	 */
	init: function(params){
		var _this = this;
		
		if (!_this.inited){
			//Minimize by default
			$('.' + _this.defaults.classNames.header)
				.filter(params.minimizedByDefault)
				.addClass(_this.defaults.classNames.minimized)
				.next()
				.hide()
			;
		}
	}
};

/**
 * jQuery.fn.mm_minimizableSections
 * @version 1.1.1 (2023-06-06)
 * 
 * @uses jQuery.ddMM.mm_minimizableSections
 * 
 * @copyright 2016
 */
$.fn.mm_minimizableSections = function(params){
	var
		_this = $.ddMM.mm_minimizableSections,
		$this = $(this)
	;
	
	$this.each(function(){
		let $this = $(this);
		
		$this
			.addClass(_this.defaults.classNames.header)
			.append('<div class="mm_minimizableSections_header_icon"><i class="far fa-minus-square"></i><i class="far fa-plus-square"></i></div>')
			.on(
				'click',
				function(){
					var $this = $(this);
					
					$this
						.next()
						.slideToggle(
							400,
							function(){
								$this.toggleClass(_this.defaults.classNames.minimized);
							}
						)
					;
				}
			)
		;
	});
	
	_this.init(params);
	
	return $this;
};
})(jQuery);

mm_minimizablesections.php
<?php
/**
 * mm_minimizableSections
 * @version 0.3 (2016-11-11)
 * 
 * @see README.md
 * 
 * @link https://code.divandesign.ru/modx/mm_minimizablesections
 * 
 * @copyright 2015–2016
 */

function mm_minimizableSections($params = []){
	//For backward compatibility
	if (
		!is_array($params) &&
		!is_object($params)
	){
		//Convert ordered list of params to named
		$params = \ddTools::orderedParamsToNamed([
			'paramsList' => func_get_args(),
			'compliance' => [
				'sections',
				'roles',
				'templates',
				'minimizedByDefault'
			]
		]);
	}
	
	//Defaults
	$params = \DDTools\ObjectTools::extend([
		'objects' => [
			(object) [
				'sections' => '',
				'minimizedByDefault' => '',
				'roles' => '',
				'templates' => '',
			],
			$params
		]
	]);
	
	if (
		!useThisRule(
			$params->roles,
			$params->templates
		)
	){
		return;
	}
	
	global $modx;
	
	$output = '';
	
	if ($modx->Event->name == 'OnDocFormPrerender'){
		$widgetDir = $modx->config['site_url'] . 'assets/plugins/managermanager/widgets/mm_minimizablesections/';
		
		$output .= includeJsCss(
			$widgetDir . 'mm_minimizableSections.css',
			'html'
		);
		$output .= includeJsCss(
			$widgetDir . 'jQuery.ddMM.mm_minimizableSections.js',
			'html',
			'jQuery.ddMM.mm_minimizableSections.js',
			'1.1.2'
		);
		
		$modx->Event->output($output);
	}elseif ($modx->Event->name == 'OnDocFormRender'){
		if ($params->sections == ''){
			$params->sections = '*';
		}
		
		$params->sections = makeArray($params->sections);
		$params->minimizedByDefault = makeArray($params->minimizedByDefault);
		
		$params->sections = array_map(
			'mm_minimizableSections_prepareSectionHeaderSelector',
			$params->sections
		);
		$params->minimizedByDefault = array_map(
			'mm_minimizableSections_prepareSectionHeaderSelector',
			$params->minimizedByDefault
		);
		
		$output .= '
//ManagerManager.mm_minimizableSections
$j("' . implode(',', $params->sections) . '", "#documentPane").mm_minimizableSections({
	minimizedByDefault: "' . implode(',', $params->minimizedByDefault) . '"
});
';
		
		$modx->Event->output($output);
	}
}

/**
 * mm_minimizableSections_prepareSectionHeaderSelector
 * @version 1.0.3 (2023-06-05)
 * 
 * @param $sectionId {string} — Section name. @required
 * 
 * @return {string}
 */
function mm_minimizableSections_prepareSectionHeaderSelector($sectionId){
	$result = '';
	
	switch ($sectionId){
		case 'access':
			$result = '#sectionAccessHeader';
		break;
		
		case '':
		//For backward compatibility
		case '*':
			$result = '.sectionHeader';
		break;
		
		default:
			$sectionId = prepareSectionId($sectionId);
			
			$result = '#' . $sectionId . '_header';
		break;
	}
	
	return $result;
}
?>